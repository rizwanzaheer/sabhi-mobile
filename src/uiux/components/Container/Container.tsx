import * as React from 'react'
import { View, ViewStyle, StyleSheet } from 'react-native'
import { withTheme } from '../../theming'
import * as Kancha from '../../types'

interface ContainerProps {
  /** Test ID used for e2e tests */
  testID?: string

  /** Width */
  w?: string | number | undefined

  /** Height */
  h?: string | number | undefined

  /** Bottom */
  b?: string | number | undefined

  /** Right */
  r?: string | number | undefined

  /** Border radius */
  br?: number | undefined

  /** Flex */
  flex?: number | undefined

  /** Pre-defined backgrounds accordign to the theme. use these where possible. */
  background?: Kancha.BrandPropOptions

  /** Temporary option to create custom color. Avoid is possible and deprecate if you can by modifying the theme */
  backgroundColor?: string

  /** Flex direction */
  flexDirection?: 'row' | 'column' | 'row-reverse' | 'column-reverse' | undefined

  /** Align items */
  alignItems?: 'flex-start' | 'flex-end' | 'center' | 'stretch' | 'baseline' | undefined

  /** Justify Content */
  justifyContent?:
    | 'flex-start'
    | 'flex-end'
    | 'center'
    | 'space-between'
    | 'space-around'
    | 'space-evenly'
    | undefined

  /** Set the bottom divider */
  dividerBottom?: boolean

  /** et the top divider */
  dividerTop?: boolean

  /** Set the bottom margin */
  margin?: number | boolean | undefined

  /** Set the bottom margin */
  marginBottom?: number | boolean | undefined

  /** Set the top margin */
  marginTop?: number | boolean | undefined

  /** Set the bottom margin */
  marginLeft?: number | boolean | undefined

  /** Set the top margin */
  marginRight?: number | boolean | undefined

  /** Set the default padding */
  padding?: number | boolean | undefined

  /** Set the default paddingHorizontal */
  paddingHorizontal?: number | boolean | undefined

  /** Set the bottom padding */
  paddingBottom?: number | boolean | undefined

  /** Set the top padding */
  paddingTop?: number | boolean | undefined

  /** Set the left padding */
  paddingLeft?: number | boolean | undefined

  /** Set the right padding */
  paddingRight?: number | boolean | undefined

  /** Enable border for debugging layouts */
  debugBorder?: boolean

  /** Enable border for debugging layouts */
  borderColor?: string

  /** Enable border for debugging layouts */
  borderWidth?: number

  /** Change debug border color */
  debugBorderColor?: string | undefined

  /** Add addionaly custom styles for a container. Use sparingly!! */
  viewStyle?: ViewStyle

  /** Disable the view from being able to detect interactions */
  disabled?: boolean

  /** A shadow level to apply */
  shadow?: number

  /** An opacity level. Use 1, 2, 3 etc  */
  opacity?: number

  theme: any
}

/**
 * `Container` is the most fundamental building block in Kancha and should be used anywhere you would use a `View`.
 * In most cases the built-in React Native `View` component is used in conjunction with some some simple layout styles like flexbox casuing a mess of styles.
 * Using a `Container` these can just be added as props for declarative layouts that are easy to read.
 *
 * ```tsx
 * <Container br={4} flex={1} alignItems={'center'}>
 *   <Text>Hey, This is a container</Text>
 * </Container>
 * ```
 */
const Container: React.FunctionComponent<ContainerProps> = props => {
  const DividerBottomStyles: ViewStyle = {
    borderBottomWidth: StyleSheet.hairlineWidth,
    borderBottomColor: props.theme.colors.primary.divider,
  }

  const DividerTopStyles: ViewStyle = {
    borderTopWidth: StyleSheet.hairlineWidth,
    borderTopColor: props.theme.colors.primary.divider,
  }

  const BaseStyles: ViewStyle = {
    /** Basic view styles */
    width: props.w,
    height: props.h,
    flex: props.flex,
    flexDirection: props.flexDirection,
    alignItems: props.alignItems,
    justifyContent: props.justifyContent,
    backgroundColor: props.background && props.theme.colors[props.background].background,
    borderRadius: props.br,
  }

  /** Conditionally spread props down to the View as styles */
  const styles: ViewStyle = {
    ...BaseStyles,
    ...(props.dividerBottom ? DividerBottomStyles : {}),
    ...(props.dividerTop ? DividerTopStyles : {}),
    ...(props.borderColor ? { borderColor: props.borderColor } : {}),
    ...(props.borderWidth ? { borderWidth: props.borderWidth } : {}),
    ...(props.backgroundColor ? { backgroundColor: props.backgroundColor } : {}),
    ...(props.debugBorder ? { borderWidth: 1, borderColor: 'red' } : {}),
    ...(props.b !== undefined ? { position: 'absolute', bottom: props.b, width: '100%' } : {}),
    ...(props.r !== undefined ? { position: 'absolute', bottom: props.r, width: '100%' } : {}),
    ...(props.disabled ? { opacity: 0.5 } : {}),
    ...(props.shadow
      ? {
          elevation: props.shadow * 10,
          shadowColor: '#ffffff',
          shadowOpacity: 0.2,
          shadowRadius: props.shadow * 5,
          backgroundColor: props.backgroundColor
            ? props.background && props.theme.colors[props.background].background
            : '#ffffff',
        }
      : {}),

    /** Margins */
    margin: typeof props.margin === 'boolean' ? props.theme.spacing.default : props.margin,
    marginBottom: typeof props.marginBottom === 'boolean' ? props.theme.spacing.default : props.marginBottom,
    marginTop: typeof props.marginTop === 'boolean' ? props.theme.spacing.default : props.marginTop,
    marginLeft: typeof props.marginLeft === 'boolean' ? props.theme.spacing.default : props.marginLeft,
    marginRight: typeof props.marginRight === 'boolean' ? props.theme.spacing.default : props.marginRight,

    /** Paddings */
    padding: typeof props.padding === 'boolean' ? props.theme.spacing.default : props.padding,
    paddingHorizontal:
      typeof props.paddingHorizontal === 'boolean' ? props.theme.spacing.default : props.paddingHorizontal,

    paddingBottom:
      typeof props.paddingBottom === 'boolean' ? props.theme.spacing.default : props.paddingBottom,
    paddingTop: typeof props.paddingTop === 'boolean' ? props.theme.spacing.default : props.paddingTop,
    paddingLeft: typeof props.paddingLeft === 'boolean' ? props.theme.spacing.default : props.paddingLeft,
    paddingRight: typeof props.paddingRight === 'boolean' ? props.theme.spacing.default : props.paddingRight,

    /** Viewstyle props will overide all options */
    ...(props.viewStyle ? { ...props.viewStyle } : {}),
  }

  return (
    <View testID={props.testID} style={styles} pointerEvents={props.disabled ? 'none' : 'auto'}>
      {props.children}
    </View>
  )
}

export default withTheme(Container)
