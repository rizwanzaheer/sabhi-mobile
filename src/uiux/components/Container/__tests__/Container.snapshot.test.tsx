import React from 'react'
import { render } from 'react-native-testing-library'

import Container from '../Container'

describe('Container', () => {
  it('renders correctly with all props', () => {
    const tree = render(
      <Container
        testID="CONTAINER_TEST_ID"
        w={100}
        h={100}
        br={5}
        flex={1}
        flexDirection={'row'}
        backgroundColor={'#ffffff'}
        background={'primary'}
        alignItems={'center'}
        justifyContent={'center'}
        dividerBottom={true}
        dividerTop={true}
        margin={true}
        marginBottom={true}
        marginLeft={true}
        marginRight={true}
        marginTop={true}
        paddingBottom={true}
        paddingTop={true}
        paddingLeft={true}
        paddingRight={true}
        debugBorder={true}
        debugBorderColor={'#000000'}
        shadow={5}
        opacity={0.6}
        borderColor={'#000000'}
        borderWidth={2}
      />,
    ).toJSON()

    expect(tree).toMatchSnapshot()
  })

  it('renders correctly with shorthand props', () => {
    const tree = render(<Container padding={true} paddingHorizontal={true} />).toJSON()

    expect(tree).toMatchSnapshot()
  })

  it('renders shadow with background', () => {
    const tree = render(<Container shadow={4} background={'primary'} />).toJSON()

    expect(tree).toMatchSnapshot()
  })

  it('renders shadow with backgroundColor', () => {
    const tree = render(<Container shadow={4} backgroundColor={'#ffffff'} />).toJSON()

    expect(tree).toMatchSnapshot()
  })

  it('renders correctly with value props', () => {
    const tree = render(
      <Container
        margin={10}
        marginBottom={10}
        marginRight={10}
        marginTop={10}
        marginLeft={10}
        paddingBottom={10}
        paddingHorizontal={10}
        paddingLeft={10}
        paddingRight={10}
        paddingTop={10}
      />,
    ).toJSON()

    expect(tree).toMatchSnapshot()
  })

  it('renders correctly with custom viewStyle', () => {
    const tree = render(<Container viewStyle={{ flex: 1, margin: 10 }} />).toJSON()

    expect(tree).toMatchSnapshot()
  })

  it('renders correctly with absolute positioning', () => {
    const tree = render(<Container b={100} r={20} />).toJSON()

    expect(tree).toMatchSnapshot()
  })

  it('renders correctly with pointer events disabled', () => {
    const tree = render(<Container disabled={true} />).toJSON()

    expect(tree).toMatchSnapshot()
  })
})
