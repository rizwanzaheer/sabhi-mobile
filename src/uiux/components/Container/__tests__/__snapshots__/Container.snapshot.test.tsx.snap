// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Container renders correctly with absolute positioning 1`] = `
<View
  pointerEvents="auto"
  style={
    Object {
      "alignItems": undefined,
      "backgroundColor": undefined,
      "borderRadius": undefined,
      "bottom": 20,
      "flex": undefined,
      "flexDirection": undefined,
      "height": undefined,
      "justifyContent": undefined,
      "margin": undefined,
      "marginBottom": undefined,
      "marginLeft": undefined,
      "marginRight": undefined,
      "marginTop": undefined,
      "padding": undefined,
      "paddingBottom": undefined,
      "paddingHorizontal": undefined,
      "paddingLeft": undefined,
      "paddingRight": undefined,
      "paddingTop": undefined,
      "position": "absolute",
      "width": "100%",
    }
  }
/>
`;

exports[`Container renders correctly with all props 1`] = `
<View
  pointerEvents="auto"
  style={
    Object {
      "alignItems": "center",
      "backgroundColor": "#FFFFFF",
      "borderBottomColor": "#CCCCCC",
      "borderBottomWidth": 0.5,
      "borderColor": "red",
      "borderRadius": 5,
      "borderTopColor": "#CCCCCC",
      "borderTopWidth": 0.5,
      "borderWidth": 1,
      "elevation": 50,
      "flex": 1,
      "flexDirection": "row",
      "height": 100,
      "justifyContent": "center",
      "margin": 15,
      "marginBottom": 15,
      "marginLeft": 15,
      "marginRight": 15,
      "marginTop": 15,
      "padding": undefined,
      "paddingBottom": 15,
      "paddingHorizontal": undefined,
      "paddingLeft": 15,
      "paddingRight": 15,
      "paddingTop": 15,
      "shadowColor": "#ffffff",
      "shadowOpacity": 0.2,
      "shadowRadius": 25,
      "width": 100,
    }
  }
  testID="CONTAINER_TEST_ID"
/>
`;

exports[`Container renders correctly with custom viewStyle 1`] = `
<View
  pointerEvents="auto"
  style={
    Object {
      "alignItems": undefined,
      "backgroundColor": undefined,
      "borderRadius": undefined,
      "flex": 1,
      "flexDirection": undefined,
      "height": undefined,
      "justifyContent": undefined,
      "margin": 10,
      "marginBottom": undefined,
      "marginLeft": undefined,
      "marginRight": undefined,
      "marginTop": undefined,
      "padding": undefined,
      "paddingBottom": undefined,
      "paddingHorizontal": undefined,
      "paddingLeft": undefined,
      "paddingRight": undefined,
      "paddingTop": undefined,
      "width": undefined,
    }
  }
/>
`;

exports[`Container renders correctly with pointer events disabled 1`] = `
<View
  pointerEvents="none"
  style={
    Object {
      "alignItems": undefined,
      "backgroundColor": undefined,
      "borderRadius": undefined,
      "flex": undefined,
      "flexDirection": undefined,
      "height": undefined,
      "justifyContent": undefined,
      "margin": undefined,
      "marginBottom": undefined,
      "marginLeft": undefined,
      "marginRight": undefined,
      "marginTop": undefined,
      "opacity": 0.5,
      "padding": undefined,
      "paddingBottom": undefined,
      "paddingHorizontal": undefined,
      "paddingLeft": undefined,
      "paddingRight": undefined,
      "paddingTop": undefined,
      "width": undefined,
    }
  }
/>
`;

exports[`Container renders correctly with shorthand props 1`] = `
<View
  pointerEvents="auto"
  style={
    Object {
      "alignItems": undefined,
      "backgroundColor": undefined,
      "borderRadius": undefined,
      "flex": undefined,
      "flexDirection": undefined,
      "height": undefined,
      "justifyContent": undefined,
      "margin": undefined,
      "marginBottom": undefined,
      "marginLeft": undefined,
      "marginRight": undefined,
      "marginTop": undefined,
      "padding": 15,
      "paddingBottom": undefined,
      "paddingHorizontal": 15,
      "paddingLeft": undefined,
      "paddingRight": undefined,
      "paddingTop": undefined,
      "width": undefined,
    }
  }
/>
`;

exports[`Container renders correctly with value props 1`] = `
<View
  pointerEvents="auto"
  style={
    Object {
      "alignItems": undefined,
      "backgroundColor": undefined,
      "borderRadius": undefined,
      "flex": undefined,
      "flexDirection": undefined,
      "height": undefined,
      "justifyContent": undefined,
      "margin": 10,
      "marginBottom": 10,
      "marginLeft": 10,
      "marginRight": 10,
      "marginTop": 10,
      "padding": undefined,
      "paddingBottom": 10,
      "paddingHorizontal": 10,
      "paddingLeft": 10,
      "paddingRight": 10,
      "paddingTop": 10,
      "width": undefined,
    }
  }
/>
`;

exports[`Container renders shadow with background 1`] = `
<View
  pointerEvents="auto"
  style={
    Object {
      "alignItems": undefined,
      "backgroundColor": "#ffffff",
      "borderRadius": undefined,
      "elevation": 40,
      "flex": undefined,
      "flexDirection": undefined,
      "height": undefined,
      "justifyContent": undefined,
      "margin": undefined,
      "marginBottom": undefined,
      "marginLeft": undefined,
      "marginRight": undefined,
      "marginTop": undefined,
      "padding": undefined,
      "paddingBottom": undefined,
      "paddingHorizontal": undefined,
      "paddingLeft": undefined,
      "paddingRight": undefined,
      "paddingTop": undefined,
      "shadowColor": "#ffffff",
      "shadowOpacity": 0.2,
      "shadowRadius": 20,
      "width": undefined,
    }
  }
/>
`;

exports[`Container renders shadow with backgroundColor 1`] = `
<View
  pointerEvents="auto"
  style={
    Object {
      "alignItems": undefined,
      "backgroundColor": undefined,
      "borderRadius": undefined,
      "elevation": 40,
      "flex": undefined,
      "flexDirection": undefined,
      "height": undefined,
      "justifyContent": undefined,
      "margin": undefined,
      "marginBottom": undefined,
      "marginLeft": undefined,
      "marginRight": undefined,
      "marginTop": undefined,
      "padding": undefined,
      "paddingBottom": undefined,
      "paddingHorizontal": undefined,
      "paddingLeft": undefined,
      "paddingRight": undefined,
      "paddingTop": undefined,
      "shadowColor": "#ffffff",
      "shadowOpacity": 0.2,
      "shadowRadius": 20,
      "width": undefined,
    }
  }
/>
`;
